logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/EphemeralMessageDeletionHandler.kt

```diff
@@ -0,0 +1,84 @@
+package com.wire.kalium.logic.feature.message
+
+import com.wire.kalium.logic.data.id.ConversationId
+import com.wire.kalium.logic.data.message.Message
+import com.wire.kalium.logic.data.message.MessageRepository
+import com.wire.kalium.logic.functional.map
+import com.wire.kalium.logic.functional.onSuccess
+import com.wire.kalium.util.KaliumDispatcher
+import com.wire.kalium.util.KaliumDispatcherImpl
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.datetime.Clock
+import kotlin.coroutines.CoroutineContext
+
+interface EphemeralMessageDeletionHandler {
+
+    fun startSelfDeletion(conversationId: ConversationId, messageId: String)
+    fun enqueuePendingSelfDeletionMessages()
+}
+
+// TODO:Mateusz: implement failure logic (needs to be discussed with a team)
+internal class EphemeralMessageDeletionHandlerImpl(
+    private val messageRepository: MessageRepository,
+    private val kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl,
+    userSessionCoroutineScope: CoroutineScope
+) : EphemeralMessageDeletionHandler, CoroutineScope by userSessionCoroutineScope {
+    override val coroutineContext: CoroutineContext
+        get() = kaliumDispatcher.default
+
+    private val onGoingSelfDeletionMessagesMutex = Mutex()
+    private val onGoingSelfDeletionMessages = mutableMapOf<Pair<ConversationId, String>, Unit>()
```

Comment:
I do not think this is going to work since it will have a key per conversation, but we want to have it grouped per messages, so conversation + message id is kind of a merged key making every message unique